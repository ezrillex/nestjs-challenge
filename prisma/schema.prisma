// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

// npx prisma migrate dev --name the_name
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Testing {
  id   String @id @default(uuid())
  test String
}

enum roles {
  customer
  manager
  admin
  anonymous // public
}

model Users {
  id                                 String     @id @default(uuid())
  first_name                         String     @db.VarChar(60)
  last_name                          String     @db.VarChar(120)
  email                              String     @unique
  password                           String     @db.Char(60) // hashed and salted password
  created_at                         DateTime   @default(now())
  role                               roles      @default(customer)
  failed_login_attempts              Int        @default(0)
  failed_login_attempts_timestamps   DateTime[] @default([])
  last_login_at                      DateTime?
  password_reset_requests            Int        @default(0)
  password_reset_requests_timestamps DateTime[] @default([])
  password_reset_token               String?
  password_last_updated              DateTime   @default(now())
}
