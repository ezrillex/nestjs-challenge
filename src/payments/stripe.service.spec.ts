import { Test, TestingModule } from '@nestjs/testing';
import { PrismaService } from '../prisma/prisma.service';
import { StripeService } from './stripe.service';
import { ConfigService } from '@nestjs/config';
import Stripe from 'stripe';
import { PaymentIntents } from '@prisma/client';

jest.mock('stripe');

describe('Stripe Service', () => {
  let service: StripeService;
  let prismaService: PrismaService;
  let configService: ConfigService;
  const stripeMock = Stripe as jest.MockedClass<typeof Stripe>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ConfigService, PrismaService, StripeService],
    }).compile();

    service = module.get<StripeService>(StripeService);
    prismaService = module.get<PrismaService>(PrismaService);
    configService = module.get<ConfigService>(ConfigService);
  });
  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should be defined', () => {
    expect(service.webhook).toBeDefined();
  });

  it('should be defined', () => {
    expect(service.createPaymentIntent).toBeDefined();
  });

  it('should be defined', () => {
    expect(service.updatePaymentIntent).toBeDefined();
  });

  it('should be defined', () => {
    expect(service.getOrderPayments).toBeDefined();
  });

  it('should be initialized with key and version, we want to break the test if a version is upgraded', () => {
    expect(stripeMock).toHaveBeenCalledWith(
      expect.stringMatching(/^sk_(test|live)_[A-Za-z0-9]+$/),
      {
        apiVersion: '2024-12-18.acacia',
      },
    );
  });

  describe('create payment intent', () => {
    it('should throw if amount is not positive', async () => {
      await expect(
        service.createPaymentIntent(
          -1,
          '8b3ae683-0626-44be-b591-9271e288388f',
          '8b3ae683-0626-44be-b591-9271e288388f',
        ),
      ).rejects.toMatchSnapshot('error amount is negative');
    });
    it('should throw if amount is zero', async () => {
      await expect(
        service.createPaymentIntent(
          0,
          '8b3ae683-0626-44be-b591-9271e288388f',
          '8b3ae683-0626-44be-b591-9271e288388f',
        ),
      ).rejects.toMatchSnapshot('error amount is zero');
    });

    it('should throw if order is not found', async () => {
      await expect(
        service.createPaymentIntent(
          30,
          '8b3ae683-0626-44be-b591-9271e288388f',
          '8b3ae683-0626-44be-b591-9271e288388f',
        ),
      ).rejects.toMatchSnapshot('error order not found');
    });

    it('should pass request to prisma intents table', async () => {
      const mockEventFunction = jest.fn(() => {
        throw new Error('mock');
      });
      stripeMock.prototype.paymentIntents = {
        create: mockEventFunction,
      } as unknown as Stripe.PaymentIntentsResource;

      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      const spy = jest
        .spyOn(prismaService.paymentIntents, 'create')
        .mockResolvedValue({} as PaymentIntents);

      await expect(
        service.createPaymentIntent(
          30,
          '8b3ae683-0626-44be-b591-9271e288388f',
          '8b3ae683-0626-44be-b591-9271e288388f',
        ),
      ).rejects.toThrow('mock');

      expect(spy.mock.calls).toMatchSnapshot('query to create intent ok');
    });

    it('should pass request to stripe service', async () => {
      const mockEventFunction = jest.fn(() => {
        throw new Error('mock');
      });
      stripeMock.prototype.paymentIntents = {
        create: mockEventFunction,
      } as unknown as Stripe.PaymentIntentsResource;

      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      jest
        .spyOn(prismaService.paymentIntents, 'create')
        .mockResolvedValue({ id: 'some id' } as PaymentIntents);

      await expect(
        service.createPaymentIntent(
          30,
          '8b3ae683-0626-44be-b591-9271e288388f',
          '8b3ae683-0626-44be-b591-9271e288388f',
        ),
      ).rejects.toThrow('mock');

      expect(mockEventFunction.mock.calls).toMatchSnapshot(
        'query to create intent stripe',
      );
    });

    it('should update prisma with response of service and return key data', async () => {
      const mockEventFunction = jest.fn(() => {
        return {
          id: 'payment intent id',
          client_secret: 'test client secret',
        };
      });
      stripeMock.prototype.paymentIntents = {
        create: mockEventFunction,
      } as unknown as Stripe.PaymentIntentsResource;

      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      jest
        .spyOn(prismaService.paymentIntents, 'create')
        .mockResolvedValue({ id: 'some id' } as PaymentIntents);

      jest.spyOn(prismaService.paymentIntents, 'update').mockResolvedValue({
        id: 'updated intent prisma id',
      } as PaymentIntents);

      await expect(
        service.createPaymentIntent(
          30,
          '8b3ae683-0626-44be-b591-9271e288388f',
          '8b3ae683-0626-44be-b591-9271e288388f',
        ),
      ).resolves.toMatchSnapshot('returns ok');

      expect(mockEventFunction.mock.calls).toMatchSnapshot(
        'query to create intent stripe',
      );
    });
  });

  describe('webhook', () => {
    it('should throw if no headers for signature', async () => {
      await expect(
        service.webhook({ headers: {} } as Request, null),
      ).rejects.toThrowErrorMatchingSnapshot(
        'missing signature in header error',
      );
    });

    it('should error if empty body', async () => {
      await expect(
        service.webhook(
          {
            headers: { 'stripe-signature': 'some-value' },
            body: null,
          } as unknown as Request,
          null,
        ),
      ).rejects.toThrowErrorMatchingSnapshot('missing body in request error');
    });

    it('should pass data from request and body to sign stripe function', async () => {
      const mockEventFunction = jest.fn(() => {
        throw new Error('mock event');
      });
      stripeMock.prototype.webhooks = {
        constructEvent: mockEventFunction,
      } as unknown as Stripe.Webhooks;

      jest.spyOn(configService, 'get').mockReturnValue('sk_test_testing_key');

      await expect(
        service.webhook(
          {
            headers: { 'stripe-signature': 'some-value' },
            body: { random: 'value' },
          } as unknown as Request,
          Buffer.from('test'),
        ),
      ).rejects.toThrow('mock event');
      expect(mockEventFunction).toHaveBeenCalledWith(
        Buffer.from('test'),
        'some-value',
        'sk_test_testing_key',
      );
    });

    it('should get data from stripe and try to form a request to prisma and throws if prisma returns nothing', async () => {
      const spy = jest
        .spyOn(prismaService.orders, 'count')
        .mockResolvedValue(0);

      const mockEventFunction = jest.fn(() => {
        return {
          data: {
            object: {
              metadata: {
                order_id: 'test_id',
                payment_id: 'test_payment_id',
              },
            },
          },
        };
      });
      stripeMock.prototype.webhooks = {
        constructEvent: mockEventFunction,
      } as unknown as Stripe.Webhooks;

      jest.spyOn(configService, 'get').mockReturnValue('sk_test_testing_key');

      await expect(
        service.webhook(
          {
            headers: { 'stripe-signature': 'some-value' },
            body: { random: 'value' },
          } as unknown as Request,
          Buffer.from('test'),
        ),
      ).rejects.toThrow('Specified order does not exist');
      expect(spy.mock.calls).toMatchSnapshot('prisma find payment query');
    });

    it('should throw if event type unkown', async () => {
      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      const mockEventFunction = jest.fn(() => {
        return {
          type: 'maracas',
          data: {
            object: {
              metadata: {
                order_id: 'test_id',
                payment_id: 'test_payment_id',
              },
            },
          },
        };
      });
      stripeMock.prototype.webhooks = {
        constructEvent: mockEventFunction,
      } as unknown as Stripe.Webhooks;

      jest.spyOn(configService, 'get').mockReturnValue('sk_test_testing_key');

      await expect(
        service.webhook(
          {
            headers: { 'stripe-signature': 'some-value' },
            body: { random: 'value' },
          } as unknown as Request,
          Buffer.from('test'),
        ),
      ).rejects.toThrow('Unknown event type.');
    });

    it('should form an update request to prisma with success and return received true for stripe', async () => {
      jest.useFakeTimers();
      jest.setSystemTime(new Date(2020, 12, 1, 10, 10, 59));
      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      const mockEventFunction = jest.fn(() => {
        return {
          type: 'payment_intent.succeeded',
          data: {
            object: {
              metadata: {
                order_id: 'test_id',
                payment_id: 'test_payment_id',
              },
            },
          },
        };
      });
      stripeMock.prototype.webhooks = {
        constructEvent: mockEventFunction,
      } as unknown as Stripe.Webhooks;

      jest.spyOn(configService, 'get').mockReturnValue('sk_test_testing_key');

      const spy = jest
        .spyOn(prismaService.orders, 'update')
        .mockResolvedValue(null);

      await expect(
        service.webhook(
          {
            headers: { 'stripe-signature': 'some-value' },
            body: { random: 'value' },
          } as unknown as Request,
          Buffer.from('test'),
        ),
      ).resolves.toEqual({
        received: true,
      });

      expect(spy.mock.calls).toMatchSnapshot('update status ok order');
    });

    it('should form an update request to prisma with failed and return received true for stripe', async () => {
      jest.useFakeTimers();
      jest.setSystemTime(new Date(2020, 12, 1, 10, 10, 59));
      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      const mockEventFunction = jest.fn(() => {
        return {
          type: 'payment_intent.payment_failed',
          data: {
            object: {
              metadata: {
                order_id: 'test_id',
                payment_id: 'test_payment_id',
              },
            },
          },
        };
      });
      stripeMock.prototype.webhooks = {
        constructEvent: mockEventFunction,
      } as unknown as Stripe.Webhooks;

      jest.spyOn(configService, 'get').mockReturnValue('sk_test_testing_key');

      const spy = jest
        .spyOn(prismaService.orders, 'update')
        .mockResolvedValue(null);

      await expect(
        service.webhook(
          {
            headers: { 'stripe-signature': 'some-value' },
            body: { random: 'value' },
          } as unknown as Request,
          Buffer.from('test'),
        ),
      ).resolves.toEqual({
        received: true,
      });

      expect(spy.mock.calls).toMatchSnapshot(
        'update status payment_failed order',
      );
    });
  });

  describe('debug method, update payment intent', () => {
    it('throws if cant find the payment intent', async () => {
      jest
        .spyOn(prismaService.paymentIntents, 'findUnique')
        .mockResolvedValue(null);

      await expect(
        service.updatePaymentIntent('payment id', 'new payment method'),
      ).rejects.toThrowErrorMatchingSnapshot('intent not found error');
    });

    it('throws if status is already completed', async () => {
      jest.spyOn(prismaService.paymentIntents, 'findUnique').mockResolvedValue({
        status: 'succeeded',
      } as PaymentIntents);

      await expect(
        service.updatePaymentIntent('payment id', 'new payment method'),
      ).rejects.toThrowErrorMatchingSnapshot('intent already completed error');
    });

    it('passes a query to update to prisma', async () => {
      jest.spyOn(prismaService.paymentIntents, 'findUnique').mockResolvedValue({
        stripe_event_id: 'test_id',
        status: 'requires_payment_method',
      } as PaymentIntents);

      const mockUpdateFunction = jest.fn();
      stripeMock.prototype.paymentIntents = {
        update: mockUpdateFunction,
      } as unknown as Stripe.PaymentIntentsResource;

      await expect(
        service.updatePaymentIntent('payment id', 'new payment method'),
      ).resolves.not.toThrow();

      expect(mockUpdateFunction.mock.calls).toMatchSnapshot(
        'sends the update payment method to stripe in query',
      );
    });
  });

  describe('get order payments', () => {
    it('throws if cant find order with user', async () => {
      await expect(
        service.getOrderPayments(
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('order user combo not found');
    });

    it('if valid queries for a list with all the payment intents associated', async () => {
      jest.spyOn(prismaService.orders, 'count').mockResolvedValue(1);

      const spy = jest
        .spyOn(prismaService.paymentIntents, 'findMany')
        .mockResolvedValue(null);

      await expect(
        service.getOrderPayments(
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
        ),
      ).resolves.not.toThrow();

      expect(spy.mock.calls).toMatchSnapshot(
        'query of payments with select fields',
      );
    });
  });
});
