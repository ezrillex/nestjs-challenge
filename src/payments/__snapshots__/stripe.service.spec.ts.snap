// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stripe Service create payment intent should pass request to prisma intents table: query to create intent ok 1`] = `
[
  [
    {
      "data": {
        "order": {
          "connect": {
            "id": "8b3ae683-0626-44be-b591-9271e288388f",
          },
        },
      },
    },
  ],
]
`;

exports[`Stripe Service create payment intent should pass request to stripe service: query to create intent stripe 1`] = `
[
  [
    {
      "amount": 30,
      "automatic_payment_methods": {
        "allow_redirects": "never",
        "enabled": true,
      },
      "currency": "usd",
      "metadata": {
        "order_id": "8b3ae683-0626-44be-b591-9271e288388f",
        "payment_id": "some id",
      },
    },
  ],
]
`;

exports[`Stripe Service create payment intent should throw if amount is not positive: error amount is negative 1`] = `[BadRequestException: Amount cant be negative or zero!]`;

exports[`Stripe Service create payment intent should throw if amount is zero: error amount is zero 1`] = `[BadRequestException: Amount cant be negative or zero!]`;

exports[`Stripe Service create payment intent should throw if order is not found: error order not found 1`] = `[NotFoundException: Order specified not found. Or order does not belong to the logged in user!]`;

exports[`Stripe Service create payment intent should update prisma with response of service and return key data: query to create intent stripe 1`] = `
[
  [
    {
      "amount": 30,
      "automatic_payment_methods": {
        "allow_redirects": "never",
        "enabled": true,
      },
      "currency": "usd",
      "metadata": {
        "order_id": "8b3ae683-0626-44be-b591-9271e288388f",
        "payment_id": "some id",
      },
    },
  ],
]
`;

exports[`Stripe Service create payment intent should update prisma with response of service and return key data: returns ok 1`] = `
{
  "client_secret": "test client secret",
  "created_at": 2021-01-12T18:12:12.000Z,
  "id": "8b3ae683-0626-44be-b591-9271e288388f",
  "order_id": "8b3ae683-0626-44be-b591-9271e288388f",
  "status": "requires_payment_method",
  "stripe_event_id": "8b3ae683-0626-44be-b591-9271e288388f",
}
`;

exports[`Stripe Service debug method, update payment intent passes a query to update to prisma: sends the update payment method to stripe in query 1`] = `
[
  [
    "test_id",
    {
      "payment_method": "new payment method",
    },
  ],
]
`;

exports[`Stripe Service debug method, update payment intent throws if cant find the payment intent: intent not found error 1`] = `"Payment intent specified not found."`;

exports[`Stripe Service debug method, update payment intent throws if status is already completed: intent already completed error 1`] = `"Payment has already succeeded, no further changes are allowed."`;

exports[`Stripe Service get order payments if valid queries for a list with all the payment intents associated: query of payments with select fields 1`] = `
[
  [
    {
      "select": {
        "created_at": true,
        "id": true,
        "order_id": true,
        "status": true,
        "stripe_event_id": true,
      },
      "where": {
        "order_id": "076b5b00-c719-40c3-a8f2-d1a11c17b75c",
      },
    },
  ],
]
`;

exports[`Stripe Service get order payments throws if cant find order with user: order user combo not found 1`] = `"Order specified not found. Or order does not belong to the logged in user!"`;

exports[`Stripe Service webhook should error if empty body: missing body in request error 1`] = `"Missing Body"`;

exports[`Stripe Service webhook should form an update request to prisma with failed and return received true for stripe: update status payment_failed order 1`] = `
[
  [
    {
      "data": {
        "IncomingPaymentWebhooks": {
          "create": {
            "data": "{"signature":"some-value","body":{"random":"value"}}",
            "processed_at": "2021-01-01T16:10:59.000Z",
          },
        },
        "PaymentIntents": {
          "update": {
            "data": {
              "status": undefined,
            },
            "where": {
              "id": "test_payment_id",
            },
          },
        },
        "paymentStatus": undefined,
      },
      "where": {
        "id": "test_id",
      },
    },
  ],
]
`;

exports[`Stripe Service webhook should form an update request to prisma with success and return received true for stripe: update status ok order 1`] = `
[
  [
    {
      "data": {
        "IncomingPaymentWebhooks": {
          "create": {
            "data": "{"signature":"some-value","body":{"random":"value"}}",
            "processed_at": "2021-01-01T16:10:59.000Z",
          },
        },
        "PaymentIntents": {
          "update": {
            "data": {
              "status": "succeeded",
            },
            "where": {
              "id": "test_payment_id",
            },
          },
        },
        "paymentStatus": "succeeded",
      },
      "where": {
        "id": "test_id",
      },
    },
  ],
]
`;

exports[`Stripe Service webhook should get data from stripe and try to form a request to prisma and throws if prisma returns nothing: prisma find payment query 1`] = `
[
  [
    {
      "where": {
        "PaymentIntents": {
          "some": {
            "id": "test_payment_id",
          },
        },
        "id": "test_id",
      },
    },
  ],
]
`;

exports[`Stripe Service webhook should throw if no headers for signature: missing signature in header error 1`] = `"Missing Stripe Signature"`;
